File: Reqnroll/Chapter18/18.4-SeededDataPattern/After/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(OrderingApiDriver orderingApiDriver, AuthenticationApiDriver authApiDriver)
{
    private int? placedOrderNo;

    [Given("the customer has placed an order containing a {string} pizza")]
    public async Task GivenTheCustomerHasPlacedAnOrderContainingAPizza(string pizzaName)
    {
        await authApiDriver
            .Login(DomainDefaults.CustomerName, DomainDefaults.Password)
            .Execute();
        var orderRequest = new PlaceOrderRequestObjectMother()
            .WithItem(name: pizzaName)
            .Build();
        var placedOrder = await orderingApiDriver.PlaceOrder(orderRequest)
            .Execute();
        placedOrderNo = placedOrder.OrderNo;
    }

    [When("the order is delivered")]
    public async Task WhenTheOrderIsDelivered()
    {
        await orderingApiDriver.DeliverOrder(placedOrderNo ?? throw new InvalidOperationException("Order not placed"))
            .Execute();
    }
}

File: CucumberJVM/Chapter21/21.2-ResourcePoolingPattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.specs.drivers.OrderingApiDriver;
import com.wimp.app.specs.support.AuthenticationContext;
import com.wimp.app.specs.support.OrderingContext;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import io.cucumber.java.en.*;

public class OrderingStepDefinitions {
    private final OrderingContext orderingContext;
    private final AuthenticationContext authenticationContext;
    private final OrderingApiDriver orderingApiDriver;

    public OrderingStepDefinitions(OrderingContext orderingContext, AuthenticationContext authenticationContext, OrderingApiDriver orderingApiDriver) {
        this.orderingContext = orderingContext;
        this.authenticationContext = authenticationContext;
        this.orderingApiDriver = orderingApiDriver;
    }

    @Given("the customer has placed an order containing a {string} pizza")
    public void theCustomerHasPlacedAnOrderContainingAPizza(String pizzaName) throws Exception {
        authenticationContext.ensureAuthenticatedCustomer();
        var placeOrderRequest = new PlaceOrderRequestObjectMother().withItem(pizzaName).build();
        var placedOrder = orderingApiDriver.placeOrder(placeOrderRequest).execute();
        orderingContext.setCurrentOrderNo(placedOrder.getOrderNo());
    }

    @When("the order is delivered")
    public void theOrderIsDelivered() throws Exception {
        orderingApiDriver.deliverOrder(orderingContext.getCurrentOrderNoVerified()).execute();
    }
}
